File: sources/roblox-lua-items-network/shared/kartEngine/KartShared/UnityCameraFollow.lua
-- Unity风格的相机跟随系统
-- 完全按照Unity的相机跟随逻辑实现，不简化任何逻辑
local UnityCameraFollow = {}
UnityCameraFollow.__index = UnityCameraFollow

local RunService = game:GetService("RunService")

-- 保存Roblox原生Vector3引用，避免命名冲突
local RobloxVector3 = Vector3

-- Unity数学库
local UnityMath = require(script.Parent.UnityMath)
local Vector3 = UnityMath.Vector3
local Quaternion = UnityMath.Quaternion

function UnityCameraFollow.new(camera, target, rigidbodyWalker)
    local self = setmetatable({}, UnityCameraFollow)
    
    -- 相机设置
    self.camera = camera or workspace.CurrentCamera
    self.target = target
    self.rigidbodyWalker = rigidbodyWalker
    
    -- Unity风格的相机参数
    self.height = 2.0
    self.distance = 10.0
    self.heightDamping = 8.0  -- 增加高度跟随速度
    self.rotationDamping = 15.0  -- 增加旋转跟随速度
    self.followSpeed = 8.0  -- 增加位置跟随速度
    self.lookAhead = 2.0
    
    -- 内部状态
    self.currentHeight = 0
    self.wantedHeight = 0
    
    -- 连接
    self.connection = nil
    
    return self
end

-- 开始相机跟随
function UnityCameraFollow:Start()
    if self.connection then
        self.connection:Disconnect()
    end
    
    self.connection = RunService.Heartbeat:Connect(function(deltaTime)
        self:UpdateCamera(deltaTime)
    end)
end

-- 停止相机跟随
function UnityCameraFollow:Stop()
    if self.connection then
        self.connection:Disconnect()
        self.connection = nil
    end
end

-- Unity风格的相机更新逻辑
function UnityCameraFollow:UpdateCamera(deltaTime)
    if not self.target then
        return
    end
    
    -- 获取目标位置和方向
    local targetPosition = self:GetTargetPosition()
    local targetForward = self:GetTargetForward()
    local targetVelocity = self:GetTargetVelocity()
    
    -- 检查获取的数据是否有效
    if not targetPosition or not targetForward then
        return
    end
    
    -- 计算相机方向 - 优先使用车头朝向，响应更快
    local cameraDirection = targetForward
    
    -- 计算期望高度
    self.wantedHeight = targetPosition.Y + self.height
    self.currentHeight = self:Lerp(self.currentHeight, self.wantedHeight, self.heightDamping * deltaTime)
    
    -- 计算相机位置（在目标后方）
    local backwardDirection = cameraDirection * -1
    local desiredPosition = targetPosition + backwardDirection * self.distance
    desiredPosition = Vector3.new(desiredPosition.X, self.currentHeight, desiredPosition.Z)
    
    -- 应用平滑跟随
    local currentPos = self:RobloxToUnityVector3(self.camera.CFrame.Position)
    local smoothedPosition = self:LerpVector3(currentPos, desiredPosition, self.followSpeed * deltaTime)
    
    -- 转换为Roblox坐标系并设置相机
    local robloxCameraPos = self:UnityToRobloxVector3(smoothedPosition)
    local robloxTargetPos = self:UnityToRobloxVector3(targetPosition)
    
    -- 确保转换成功（应该是Roblox Vector3 userdata类型）
    if not robloxCameraPos or not robloxTargetPos then
        warn("相机位置转换失败!")
        return
    end
    
    -- 使用简单的CFrame.lookAt方式，如果失败则使用备用方案
    local success, cameraCFrame = pcall(function()
        return CFrame.lookAt(robloxCameraPos, robloxTargetPos)
    end)
    
    if success then
        self.camera.CFrame = cameraCFrame
    else
        -- 备用方案：手动构造CFrame
        warn("CFrame.lookAt失败，使用备用方案")
        local lookDirection = (robloxTargetPos - robloxCameraPos).Unit
        self.camera.CFrame = CFrame.new(robloxCameraPos, robloxCameraPos + lookDirection)
    end
end

-- 获取目标位置
function UnityCameraFollow:GetTargetPosition()
    if not self.target then
        return nil
    end
    
    if self.target.ClassName == "Model" then
        local primaryPart = self.target.PrimaryPart or self.target:FindFirstChildOfClass("BasePart")
        if primaryPart then
            return self:RobloxToUnityVector3(primaryPart.Position)
        else
            return Vector3.zero
        end
    else
        return self:RobloxToUnityVector3(self.target.Position)
    end
end

-- 获取目标前进方向
function UnityCameraFollow:GetTargetForward()
    if not self.target then
        return nil
    end
    
    if self.target.ClassName == "Model" then
        local primaryPart = self.target.PrimaryPart or self.target:FindFirstChildOfClass("BasePart")
        if primaryPart then
            -- 使用 -LookVector 作为前进方向，与Unity逻辑保持一致
            local lookVec = primaryPart.CFrame.LookVector
            return self:RobloxToUnityVector3(RobloxVector3.new(-lookVec.X, -lookVec.Y, -lookVec.Z))
        else
            return Vector3.new(0, 0, -1)
        end
    else
        -- 使用 -LookVector 作为前进方向，与Unity逻辑保持一致
        local lookVec = self.target.CFrame.LookVector
        return self:RobloxToUnityVector3(RobloxVector3.new(-lookVec.X, -lookVec.Y, -lookVec.Z))
    end
end

-- 获取目标速度
function UnityCameraFollow:GetTargetVelocity()
    if self.rigidbodyWalker and self.rigidbodyWalker.goPlayKart_ then
        local kartVelocity = self.rigidbodyWalker.goPlayKart_.m_KartRealVelocity
        if kartVelocity then
            return kartVelocity
        end
    end
    return Vector3.zero
end

-- 坐标系转换函数
function UnityCameraFollow:RobloxToUnityVector3(robloxVec3)
    if not robloxVec3 then
        return Vector3.zero
    end
    return Vector3.new(robloxVec3.X, robloxVec3.Y, robloxVec3.Z)
end

function UnityCameraFollow:UnityToRobloxVector3(unityVec3)
    if not unityVec3 then
        return RobloxVector3.new(0, 0, 0)
    end
    
    -- 确保访问的是正确的属性
    local x = unityVec3.X or 0
    local y = unityVec3.Y or 0
    local z = unityVec3.Z or 0
    
    return RobloxVector3.new(x, y, z)
end

-- 设置相机参数
function UnityCameraFollow:SetDistance(distance)
    self.distance = distance
end

function UnityCameraFollow:SetHeight(height)
    self.height = height
end

function UnityCameraFollow:SetDamping(heightDamping, rotationDamping)
    self.heightDamping = heightDamping
    self.rotationDamping = rotationDamping
end

function UnityCameraFollow:SetTarget(target)
    self.target = target
end

-- 辅助数学函数
function UnityCameraFollow:Lerp(a, b, t)
    local clampedT = math.min(math.max(t, 0), 1)
    return a + (b - a) * clampedT
end

function UnityCameraFollow:LerpVector3(a, b, t)
    if not a or not b then
        return a or b or Vector3.zero
    end
    
    local clampedT = math.min(math.max(t, 0), 1)
    return Vector3.new(
        a.X + (b.X - a.X) * clampedT,
        a.Y + (b.Y - a.Y) * clampedT,
        a.Z + (b.Z - a.Z) * clampedT
    )
end

-- 清理资源
function UnityCameraFollow:Destroy()
    self:Stop()
end

return UnityCameraFollow

File: src/shared/kartEngine/KartShared/UnityCameraFollow.ts
import { RunService, Workspace } from "@rbxts/services";
import { UnityVector3, Mathf } from "./UnityMath";

interface GoPlayKart {
    m_KartRealVelocity?: UnityVector3;
}

export interface RigidbodyWalker {
    goPlayKart_?: GoPlayKart;
}

export class UnityCameraFollow {
    // 相机设置
    private camera: Camera;
    private target?: Model | BasePart;
    private rigidbodyWalker?: RigidbodyWalker;
    
    // Unity风格的相机参数
    private height: number = 2.0;
    private distance: number = 10.0;
    private heightDamping: number = 8.0;  // 增加高度跟随速度
    private rotationDamping: number = 15.0;  // 增加旋转跟随速度
    private followSpeed: number = 8.0;  // 增加位置跟随速度
    private lookAhead: number = 2.0;
    
    // 内部状态
    private currentHeight: number = 0;
    private wantedHeight: number = 0;
    
    // 连接
    private connection?: RBXScriptConnection;

    constructor(camera?: Camera, target?: Model | BasePart, rigidbodyWalker?: RigidbodyWalker) {
        this.camera = camera || (Workspace.CurrentCamera as Camera);
        this.target = target;
        this.rigidbodyWalker = rigidbodyWalker;
    }

    // 开始相机跟随
    Start(): void {
        if (this.connection) {
            this.connection.Disconnect();
        }
        
        this.connection = RunService.Heartbeat.Connect((deltaTime) => {
            this.UpdateCamera(deltaTime);
        });
    }

    // 停止相机跟随
    Stop(): void {
        if (this.connection) {
            this.connection.Disconnect();
            this.connection = undefined;
        }
    }

    // Unity风格的相机更新逻辑
    private UpdateCamera(deltaTime: number): void {
        if (!this.target) {
            return;
        }
        
        // 获取目标位置和方向?        
        const targetPosition = this.GetTargetPosition();
        const targetForward = this.GetTargetForward();
        const targetVelocity = this.GetTargetVelocity();
        
        // 检查获取的数据是否有效
        if (!targetPosition || !targetForward) {
            return;
        }
        
        // 计算相机方向 - 优先使用车头朝向，响应更快
        const cameraDirection = targetForward;
        
        // 计算期望高度
        this.wantedHeight = targetPosition.Y + this.height;
        this.currentHeight = this.Lerp(this.currentHeight, this.wantedHeight, this.heightDamping * deltaTime);
        
        // 计算相机位置（在目标后方）
        const backwardDirection = new UnityVector3(-cameraDirection.X, -cameraDirection.Y, -cameraDirection.Z);
        let desiredPosition = new UnityVector3(
            targetPosition.X + backwardDirection.X * this.distance,
            targetPosition.Y + backwardDirection.Y * this.distance,
            targetPosition.Z + backwardDirection.Z * this.distance
        );
        desiredPosition = new UnityVector3(desiredPosition.X, this.currentHeight, desiredPosition.Z);
        
        // 应用平滑跟随
        const currentPos = this.RobloxToUnityVector3(this.camera.CFrame.Position);
        const smoothedPosition = this.LerpVector3(currentPos, desiredPosition, this.followSpeed * deltaTime);
        
        // 转换为Roblox坐标系并设置相机
        const robloxCameraPos = this.UnityToRobloxVector3(smoothedPosition);
        const robloxTargetPos = this.UnityToRobloxVector3(targetPosition);
        
        // 确保转换成功（应该是Roblox Vector3 userdata类型）
        if (!robloxCameraPos || !robloxTargetPos) {
            warn("相机位置转换失败!");
            return;
        }
        
        // 使用简单的CFrame.lookAt方式，如果失败则使用备用方案
        const [success, cameraCFrame] = pcall(() => {
            return CFrame.lookAt(robloxCameraPos, robloxTargetPos);
        });
        
        if (success) {
            this.camera.CFrame = cameraCFrame as CFrame;
        } else {
            // 备用方案：手动构造CFrame
            warn("CFrame.lookAt失败，使用备用方案");
            const lookDirection = robloxTargetPos.sub(robloxCameraPos).Unit;
            this.camera.CFrame = new CFrame(robloxCameraPos, robloxCameraPos.add(lookDirection));
        }
    }

    // 获取目标位置
    private GetTargetPosition(): UnityVector3 | undefined {
        if (!this.target) {
            return undefined;
        }
        
        if (this.target.IsA("Model")) {
            const primaryPart = (this.target as Model).PrimaryPart || (this.target as Model).FindFirstChildOfClass("BasePart");
            if (primaryPart) {
                return this.RobloxToUnityVector3(primaryPart.Position);
            } else {
                return UnityVector3.zero;
            }
        } else {
            return this.RobloxToUnityVector3((this.target as BasePart).Position);
        }
    }

    // 获取目标前进方向
    private GetTargetForward(): UnityVector3 | undefined {
        if (!this.target) {
            return undefined;
        }
        
        if (this.target.IsA("Model")) {
            const primaryPart = (this.target as Model).PrimaryPart || (this.target as Model).FindFirstChildOfClass("BasePart");
            if (primaryPart) {
                // 使用 -LookVector 作为前进方向，与Unity逻辑保持一致
                const lookVec = primaryPart.CFrame.LookVector;
                return this.RobloxToUnityVector3(new Vector3(-lookVec.X, -lookVec.Y, -lookVec.Z));
            } else {
                return new UnityVector3(0, 0, -1);
            }
        } else {
            // 使用 -LookVector 作为前进方向，与Unity逻辑保持一致
            const lookVec = (this.target as BasePart).CFrame.LookVector;
            return this.RobloxToUnityVector3(new Vector3(-lookVec.X, -lookVec.Y, -lookVec.Z));
        }
    }

    // 获取目标速度
    private GetTargetVelocity(): UnityVector3 {
        if (this.rigidbodyWalker && this.rigidbodyWalker.goPlayKart_) {
            const kartVelocity = this.rigidbodyWalker.goPlayKart_.m_KartRealVelocity;
            if (kartVelocity) {
                return kartVelocity;
            }
        }
        return UnityVector3.zero;
    }

    // 坐标系转换函数
    private RobloxToUnityVector3(robloxVec3: Vector3): UnityVector3 {
        if (!robloxVec3) {
            return UnityVector3.zero;
        }
        // 从 Roblox Vector3 转换为 Unity Vector3 实例
        return new UnityVector3(robloxVec3.X, robloxVec3.Y, robloxVec3.Z);
    }

    private UnityToRobloxVector3(unityVec3: UnityVector3): Vector3 {
        if (!unityVec3) {
            return new Vector3(0, 0, 0);
        }
        
        // 从 Unity Vector3 转换为 Roblox Vector3
        return new Vector3(unityVec3.X, unityVec3.Y, unityVec3.Z);
    }

    // 设置相机参数
    SetDistance(distance: number): void {
        this.distance = distance;
    }

    SetHeight(height: number): void {
        this.height = height;
    }

    SetDamping(heightDamping: number, rotationDamping: number): void {
        this.heightDamping = heightDamping;
        this.rotationDamping = rotationDamping;
    }

    SetTarget(target: Model | BasePart): void {
        this.target = target;
    }

    // 辅助数学函数
    private Lerp(a: number, b: number, t: number): number {
        const clampedT = math.min(math.max(t, 0), 1);
        return a + (b - a) * clampedT;
    }

    private LerpVector3(a: UnityVector3, b: UnityVector3, t: number): UnityVector3 {
        if (!a || !b) {
            return a || b || UnityVector3.zero;
        }
        
        const clampedT = math.min(math.max(t, 0), 1);
        return new UnityVector3(
            a.X + (b.X - a.X) * clampedT,
            a.Y + (b.Y - a.Y) * clampedT,
            a.Z + (b.Z - a.Z) * clampedT
        );
    }

    // 清理资源
    Destroy(): void {
        this.Stop();
    }
}
